test/Sign.Core.Test/ServiceProviderFactoryTests.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE.txt file in the project root for more information.
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace Sign.Core.Test
+{
+    public class ServiceProviderFactoryTests
+    {
+        [Fact]
+        public void AddService_ServicesIsNull_Throws()
+        {
+            var factory = new ServiceProviderFactory();
+
+            ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                () => factory.AddServices(null!));
+
+            Assert.Equal("addServices", exception.ParamName);
+        }
+
+        [Fact]
+        public void AddServices_WhenServicesAreNotAlreadyPresent_AddsServices()
+        {
+            var factory = new ServiceProviderFactory();
+            factory.AddServices(services => services.AddSingleton<ITestService, TestService>());
+            IServiceProvider serviceProvider = factory.Create();
+            Assert.NotNull(serviceProvider.GetRequiredService<ITestService>());
+        }
+
+        [Fact]
+        public void AddServices_WhenServiceIsAlreadyPresent_AddsOnlyNewService()
+        {
+            var factory = new ServiceProviderFactory();
+            factory.AddServices(services => services.AddSingleton<ITestService, TestService>());
+            IServiceProvider serviceProvider = factory.Create(addServices: services => services.AddSingleton<ITestService2, TestService2>());
+            Assert.NotNull(serviceProvider.GetRequiredService<ITestService>());
```

Comment:
What happens if `ITestService` is added again after it was previously added?

Reply:
@erdembayar, good question.  Dependency injection (DI) allows you do this:

```C#
services.AddSingleton<IService, Implementation0>();
services.AddSingleton<IService, Implementation1>();
```

The last one wins, so `GetRequiredService<IService>()` willreturn `Implementation1` not `Implementation0`.

DI also permits this, which is what this test is doing:

```C#
services.AddSingleton<IService, Implementation0>();
services.AddSingleton<IService, Implementation0>();
```

`GetRequiredService<IService>()` will return an instance of `Implementation0`.

I updated tests to cover 2 distinct scenarios.